feat: native typescript lambda function code

thantos opened the pull request:
Closes #153 

Depends on #109 and #108 

Function closure serialization!

- [x] Basic Function serialization
- [x] Add constructor to Function to pass in a closure, scope, and id
- [x] ~~Exhaustive~~ tests
- [x] Support embedded client calls
   - [x] Step Functions
   - [x] Lambda
   - ~~App Sync~~ include in #119
   - ~~Api Gateway Endpoints~~ Not implemented - include in #23
   - [x] Event Bridge
   - [x] Dynamo
      - Only within `$AWS.DynamoDB.*` see #132
- [x] Support arbitrary token data as env variables
- [x] Split localstack integration into a separate PR #109 
- [x] AppSync integration changes #113 
- Dynamically simplify native integration references #122 

```ts
new Function(this, 'myFunction', async (event) => {
   return Object.keys(event).length;
})
```

sam-goodwin commented on src/declaration.ts:
```diff
@@ -37,6 +37,31 @@ export class FunctionDecl<F extends AnyFunction = AnyFunction> extends BaseDecl<
+/**
+ * Instead of serializing this function to AST, this function has been hoisted to a module export with a random name.
+ */
+export class HoistedFunctionDecl<
```
How will we determine the dependencies and auto wire environment variables and iam policies? Walk the FunctionDecl like we do for the functionless integrations?

thantos replied:
Ohh, ha, this needs to be cleaned up, from before using Pulumi.

thantos replied:
There are a few options here, once is to use the serialize function like Cloudy does, another is to use codegen. I will figure this out today.

sam-goodwin commented:
Do we want to pull out pulumi's closure serializer into its own package to avoid the massive dependency ?

thantos commented:
> Do we want to pull out pulumi's closure serializer into its own package to avoid the massive dependency ?

Its just a build time dependency, does it matter?

I'd like to write our own in the longer term.

sam-goodwin commented:
Can do it later, sure. But I also noticed it has a dependency on their concept of a Resource for logging. I was wondering if that mattered at all.

thantos commented:
> Can do it later, sure. But I also noticed it has a dependency on their concept of a Resource for logging. I was wondering if that mattered at all.

Looking into this. I am not a fan of copying/forking libraries, especially complex ones. But its not OK if there is any Pulumi specific solutions.

sam-goodwin commented on .projenrc.js:
```diff
@@ -1,4 +1,7 @@
 const { typescript } = require("projen");
+
+const MIN_CDK_VERSION = "2.20.0";
```
Do we need to specify the version or can we just use default. Projen defaults to latest.

thantos replied:
As a library, we should pin to the version we suggest in the peer deps. I was having trouble keeping them all the same version as I added new CDK deps, the latest would be installed, but the older ones were not being upgrading, existentially the one in alpha. Having this constant made it easier to keep them the same version.

sam-goodwin commented on README.md:
````diff
@@ -558,6 +559,31 @@ bus
+When using `.synth()` programmatically, use the provided `asyncSynth` method to wrap your app.
+
+```ts
+const cloudAssembly = asyncSynth(app, options);
````
Maybe show a test case for clarity. Also, should this be `await asyncSynth`?

thantos replied:
Not sure if I want to show a full test, actually deploying the template is verbose. Will link to an example.

sam-goodwin commented on src/declaration.ts:
```diff
@@ -37,6 +37,31 @@ export class FunctionDecl<F extends AnyFunction = AnyFunction> extends BaseDecl<
+> extends BaseDecl<"NativeFunctionDecl", undefined> {
+  readonly _functionBrand?: F;
+  constructor(
+    readonly parameters: ParameterDecl[],
```
Why don't we also include the `FunctionDecl`? We can use it to auto-wire IAM permissions and environment variables.

thantos replied:
I had a feeling of need it eventually, but wasn't sure. This is probably where we'd end up.

thantos replied:
Current implementation extracts the integrations during the native expr parsing. 

I question the need to output the entire `FunctionDecl` and risking any compilation failures when we don't make use of most of the output. 

Going to leave this as is for now.

sam-goodwin commented on src/function.ts:
```diff
@@ -30,11 +38,50 @@ export type AnyLambda = Function<any, any>;
   readonly __functionBrand: ConditionalFunction<P, O>;
 
-  constructor(readonly resource: aws_lambda.IFunction) {
+  constructor(
```
We should add thorough code documentation with examples for each variant.

thantos replied:
In a future PR, we should import for the existing case instead of using the constructor.

```ts
new Function(this, id, async () => {})

Function.fromFunction(new aws_lambda.Function(...));
```

sam-goodwin commented on src/function.ts:
```diff
@@ -80,3 +127,53 @@ export interface Function<P, O> {
+    super();
+  }
+
+  bind(scope: Construct): aws_lambda.CodeConfig {
```
nit: `public bind`

sam-goodwin commented on src/function.ts:
```diff
@@ -30,11 +38,50 @@ export type AnyLambda = Function<any, any>;
+  public static readonly FunctionlessType = "Function";
+  readonly resource: aws_lambda.IFunction;
+
+  static promises: Promise<any>[] = [];
```
Should we make sure this is a global singleton? Or do we want to rely on the `functionless` module being a singleton because of `peerDeps`? 

My preferred method is with a symbol:

```ts
const PromisesSymbol = Symbol.for("functionless.Function.promises");

public static readonly promises = (global[PromisesSymbol] = global[PromisesSymbol] ?? []);
```

thantos replied:
Good idea.

sam-goodwin commented on src/function.ts:
```diff
@@ -80,3 +127,53 @@ export interface Function<P, O> {
+
+    const codeConfig = asset.bind(scope);
+
+    funcResource.code = {
```
Let's give a link to cloudy as a thankyou

thantos commented:
Current concerns with using Pulumi:
1. Customization - Pulumi is not a hookable solution. It is possible to use the `serialize` function to ignore variables, but not rewrite them. Best you can do is ignore them.
2. Ignoring nodes is all or nothing - In `serialize`, I can ignore a node like all CDK Constructs, but I cannot get a property off of that object to use later.
3. Serialize doesn't return top level nodes in the closure - `(event) => event` never sends the `event` object to the serialize function.
4. Logging - There is a logging Resource parameter, which is Pulumi's `Resource`, still not sure the impact of this yet.

thantos commented:
Caveat: References to functions in namespaces should be simplified to keep the resulting native function bundle small.

### Impacts:
1. Bundle Size
2. Invalid Bundles during tests

### Example

```ts
new Function(..., () => {
   $AWS.EventBridge.putEvents(...);
})
```

Should be

```ts
const putEvents = $AWS.EventBridge;
new Function(..., () => {
   putEvents(...);
})
```

### Reason

1. Pulumi will serialize the entirety of $AWS and all sub-variables when using the `$AWS.EventBridge.putEvents` form. 
   1. Using the referenced from has reduced bundle sizes by over 10x
2. Invalid function are created, specifically during tests, when Pulumi tries to include code generated by jest for global functions.
   1. The referenced form seems less likely to have this issue.

### Mitigation
1. Users can follow the pattern above
2. Support Automatic Reference of native integrations (future PR: )
3. Consider avoiding integrations in keep objects? `import { awsEventBridgePutEvents } from 'functionless'` instead of `import $AWS from 'functioness'` ?

sam-goodwin commented on package.json:
```diff
@@ -31,7 +31,7 @@
     "@types/jest": "^27.5.0",
     "@types/minimatch": "^3.0.5",
-    "@types/node": "^14",
+    "@types/node": "^12",
```
hmm what's going on here? Node 12 has recently reached end of life. 

https://endoflife.date/nodejs

thantos replied:
Hmm, it was upgraded recently, downgraded during the merge from live, and restored on my commit from a while ago, running yarn default. Should be fine now.

sam-goodwin commented on src/compile.ts:
```diff
@@ -1251,6 +1204,47 @@ export function compile(
         return;
       }
 
+      function getOutOfScopeValueNode(
```
can you add a quick comment with some examples? What is this being used for?

sam-goodwin commented on src/compile.ts:
```diff
@@ -1251,6 +1204,47 @@ export function compile(
+      }
+
+      function isSymbolOutOfScope(symbol: ts.Symbol, scope: ts.Node): boolean {
+        if (symbol) {
+          if (symbol.valueDeclaration) {
```
nit: `if (symbol?.valueDeclaration)`

sam-goodwin commented on test/function.localstack.test.ts:
```diff
@@ -239,32 +239,54 @@ localstackTestSuite("functionStack", (testResource, _stack, app) => {
+        });
+      });
+
+      return { outputs: { function: func.resource.functionName } };
```
nit: forgetting prettier again?

thantos replied:
prettier runs on save and on commit. Seems like it is fine with the format. 

![image](https://user-images.githubusercontent.com/289213/167028894-bba010bb-a06d-4a83-8ac2-b4d96e242c15.png)

SAVE

![image](https://user-images.githubusercontent.com/289213/167028906-8bed7eb3-be37-4b1a-8781-0f1f2edb8861.png)


![image](https://user-images.githubusercontent.com/289213/167028927-17483a7e-cc18-484e-8d0a-08d344e7c6b8.png)


SAVE


![image](https://user-images.githubusercontent.com/289213/167028954-b4995560-8f52-4be6-8d15-9cb35b84f0c0.png)


![image](https://user-images.githubusercontent.com/289213/167029000-1d8df662-a7bf-4db6-8daa-a2ee6568c620.png)

SAVE

![image](https://user-images.githubusercontent.com/289213/167029013-8fb48fe2-76b8-4937-a6f9-1de77c1d4078.png)

thantos replied:
Is there something specific about the single form you don't like? Should the prettier rules be updated to reflect that?

sam-goodwin replied:
No worries, just looked like it wasn't formatted because everything was on one line.

sam-goodwin commented on test/function.localstack.test.ts:
```diff
@@ -239,32 +239,54 @@ localstackTestSuite("functionStack", (testResource, _stack, app) => {
     }
   );
 
+  // Function serialization breaks when assigning an integration/construct to a variable in the closure.
+  // TODO: what should happen here?
+  testResource.skip(
+    "Call Lambda AWS SDK put event to bus with in closure reference",
+    (parent) => {
+      const bus = new EventBus<EventBusRuleInput>(parent, "bus");
+      const func = new Function(parent, "function", async () => {
+        const busbus = bus;
+        busbus({
+          "detail-type": "anyDetail",
+          source: "anySource",
+          detail: {},
+        });
+      });
```
Why does it fail? Seems fine to me - we're just referencing a value.

thantos replied:
Pulumi tries to serialize the entire construct and finds some native functions it can't serialize.

sam-goodwin replied:
We are going to need to fork pulumi?

thantos replied:
Maybe? I wanted to build up more test cases and use that to figure out what it would take to build from scratch.

sam-goodwin commented on src/aws.ts:
```diff
@@ -361,4 +361,40 @@ export namespace $AWS {
+        bootstrap: () => {},
+        /* istanbul ignore next */
+        preWarm: (prewarmContext: NativePreWarmContext) => {
+          /* istanbul ignore next */
```
what is this instanbul thing?

thantos replied:
Trying to reduce the code coverage code injected into the function closures during tests, but it has not been successful. I'll probably just remove those. The code coverage code makes up a majority of the serialized functions. Mostly just a problem for us. I also though they may be contributing to issues I was having, but I don't think they are. Just another example of where we need control.

sam-goodwin commented on src/compile.ts:
```diff
@@ -233,6 +251,18 @@ export function compile(
         );
       }
 
+      // function isEventBusPutCall(node: ts.Node): node is ts.CallExpression {
+      //   return ts.isCallExpression(node) && isEventBus(node.expression);
+      // }
+
+      // function isFunctionCall(node: ts.Node): node is ts.CallExpression {
+      //   return ts.isCallExpression(node) && isFunction(node.expression);
+      // }
+
+      // function isFunction(node: ts.Node) {
+      //   return isFunctionlessClassOfKind(node, Function.FunctionlessType);
+      // }
```
remove?

sam-goodwin commented on src/compile.ts:
```diff
@@ -269,6 +299,41 @@ export function compile(
+        );
+      }
+
+      // function isCDKConstruct(type: ts.Type): boolean {
+      //   const constructImport = imports["constructs"];
+      //   const constructSymbol = constructImport
+      //     ? checker.getSymbolAtLocation(constructImport.moduleSpecifier)
+      //     : undefined;
+      //   const constructExports = constructSymbol
+      //     ? checker.getExportsOfModule(constructSymbol)
+      //     : undefined;
+      //   const constructExportSymbol = constructExports?.find(
+      //     (s) => s.name === "Construct"
+      //   );
+
+      //   return (
+      //     (!constructExportSymbol ||
+      //       type.symbol === constructExportSymbol ||
+      //       type.getBaseTypes()?.some((t) => isCDKConstruct(t))) ??
+      //     false
+      //   );
+      // }
```
remove?

sam-goodwin commented on src/compile.ts:
```diff
@@ -407,6 +472,291 @@ export function compile(
+      /**
+       * Native Functions do not allow the creation of resources, CDK constructs or functionless.
+       */
+      function toNativeFunction(
```
Provide a comment that shows an example of how the code is transformed. I'm particularly interested in preWarmContext.

thantos replied:
Done

sam-goodwin commented on src/compile.ts:
```diff
@@ -407,6 +472,291 @@ export function compile(
+            );
+
+            if (!outOfScopeIntegrationReference) {
+              throw Error(
+                "Integration Variable must be defined out of scope: " +
+                  node.expression.getText()
+              );
+            }
```
Why?

thantos replied:
```ts
              // integration variables can be CDK constructs which will fail serialization.
              // TODO: can we relax this be determining if a integration reference will fail?
```

sam-goodwin commented on src/compile.ts:
```diff
@@ -407,6 +472,291 @@ export function compile(
+        } else if (ts.isNewExpression(node)) {
+          const newType = checker.getTypeAtLocation(node);
+          // cannot create new resources in native runtime code.
+          const functionlessKind = getFunctionlessTypeKind(newType);
+          if (getFunctionlessTypeKind(newType)) {
```
Should this be expanded to any class that extends Construct?

thantos replied:
That is what that code did below, but I was having a hard time determining if something was a Contract or not.

sam-goodwin replied:
Could we use structural typing heuristics? Or the fully qualified name of the type?

thantos replied:
Fully qualified in typescript is the path, which is different on every system, right? I wasn't able to figure out how to make good use of that. Have any examples?

thantos replied:
Not sure why I didn't notice this before, will give it a try.

https://github.com/aws/aws-cdk/blob/master/packages/@aws-cdk/core/lib/construct-compat.ts#L21

thantos replied:
Got it...


```ts
      /**
       * Heuristically evaluate the fqn of a symbol to be in a module and of a type name.
       *
       * /somePath/node_modules/{module}/somePath.{typeName}
       *
       * isInstanceOf(typeSymbol, "constructs", "Construct")
       * ex: /home/sussmans/functionless/node_modules/constructs/lib/index.js
       */
      function isInstanceOf(
        symbol: ts.Symbol,
        module: string,
        typeName: string
      ) {
        const find = /.*\/node_modules\/([^\/]*)\/.*\.(.*)$/g.exec(
          checker.getFullyQualifiedName(symbol)
        );

        const [_, mod, type] = find ?? [];

        return mod === module && type === typeName;
      }

      function isCDKConstruct(type: ts.Type): boolean {
        const typeSymbol = type.getSymbol();

        return (
          ((typeSymbol &&
            isInstanceOf(typeSymbol, "constructs", "Construct")) ||
            type.getBaseTypes()?.some((t) => isCDKConstruct(t))) ??
          false
        );
      }
```

sam-goodwin commented on src/compile.ts:
```diff
@@ -407,6 +472,291 @@ export function compile(
+              `Cannot initialize new resources in a native function, found ${functionlessKind}.`
+            );
+          }
+          // TODO: check for CDK constructs
+          // const type = checker.getTypeAtLocation(node.expression);
+          // if (isCDKConstruct(type)) {
```
Ah I see. What's the hold up here? Not sure how to determine the type?

sam-goodwin commented on src/compile.ts:
```diff
@@ -407,6 +472,291 @@ export function compile(
+          if (!child) {
+            return node;
+          } else if (Array.isArray(child)) {
+            // if multiple nodes are returned, assume the last one is the value to assign and the rest should be injected as siblings.
+            const [last, ...rest] = child.reverse();
```
I don't understand?

sam-goodwin replied:
Need comments with examples.

thantos replied:
The compiler supports returning an array of nodes, but only when the parent supports multiple nodes. I needed to inject siblings into the context. If the child returns multiple nodes and the parent cannot handle multiple nodes (ex: `VariableAssignment`), they get passed up the tree until the parent can support multiple nodes (ex: `ExpressionStatement`)

For example, if I wanted to create a client that a line needs, I wanted to return the client along with the transformed node.

```ts
bus()
```

Compiled to

```ts
const EB = new EventBridge()
await EB.putTargets()
```

Howerever... the one use case I had for this is now supported using the prewarm context, so I don't need this anymore!

sam-goodwin commented on src/compile.ts:
```diff
@@ -407,6 +472,291 @@ export function compile(
+            );
+          }
+        } else if (ts.isVariableDeclarationList(node)) {
+          // bubble any extra nodes to the top of the variable declaration list.
```
why? I don't understand.

thantos replied:
See above, will remove anyways, but it was fun... is there a better way to do this?
https://github.com/sam-goodwin/functionless/pull/103#discussion_r867284322

sam-goodwin commented on src/compile.ts:
```diff
@@ -407,6 +472,291 @@ export function compile(
+              node.name,
+              node.exclamationToken,
+              node.type,
+              child.length > 0
+                ? (child[child.length - 1] as ts.Expression)
+                : undefined
+            );
+            return [
+              ...(child.length > 1 ? child.slice(0, child.length - 1) : []),
```
Wha tis going on here? Why do we need to rearrange things like this?

sam-goodwin commented on src/event-bridge/event-bus.ts:
```diff
@@ -120,9 +124,37 @@ abstract class EventBusBase<E extends EventBusRuleInput>
+      bootstrap: (context: Function<any, any>) => {
+        this.bus.grantPutEventsTo(context.resource);
+      },
+      preWarm: (prewarmContext: NativePreWarmContext) => {
+        prewarmContext.eventBridge();
```
hmm, this looks interesting. Does NativePreWarmContext have all AWS SDK?

sam-goodwin replied:
Any reason why we don't have `preWarm` instantiate any clients it needs and return them? `prewarmContext` can provide a cache which can be used to avoid duplicate clients?

thantos replied:
thats what it does, I've just added helpers for common clients. you can also add whatever.

```ts
preWarmConext.getOrInit({  key: "anything", init: () => "hi" });
preWarmConext.get("anything") // "hi"

preWarmConext.getOrInit(PrewarmClients.LAMBDA);
preWarmConext.get(PrewarmClients.LAMBDA.key) // lambda client
```

thantos replied:
the helpers correctly apply any props and ensure that all consumers use the same implementation.

```ts
  LAMBDA: {
    key: "LAMBDA",
    init: (key, props) => new Lambda(props?.clientConfigRetriever?.(key)),
  },
```

sam-goodwin commented on src/event-bridge/event-bus.ts:
```diff
@@ -120,9 +124,37 @@ abstract class EventBusBase<E extends EventBusRuleInput>
+    // Closure event bus base
+    const eventBusName = this.eventBusName;
+    this.native = <NativeIntegration<EventBusBase<E>>>{
+      bootstrap: (context: Function<any, any>) => {
```
should this be called `configure` or `synth`? `bootstrap` may be a bit vague/ambiguous. Although, funnily enough, it's exactly what I called it in punchcard haha

thantos replied:
`bind`?

sam-goodwin replied:
Like it.

sam-goodwin commented on src/function.ts:
```diff
@@ -1,43 +1,94 @@
+       * Code that runs once per lambda invocation
+       * The pre-warm client supports singleton invocation of clients (or other logic) across all integrations in the caller function.
+       */
+      preWarm: (preWarmContext: NativePreWarmContext) => {
+        preWarmContext.lambda();
+      },
```
nit: place above `call`

sam-goodwin commented on src/function.ts:
```diff
@@ -1,43 +1,94 @@
+       */
+      call: async (args, prewarmContext) => {
+        const [payload] = args;
+        const lambdaClient = prewarmContext.lambda();
```
does this prewarmContext properly tree-shake? If i only use `lambda()`, will it also serialize other client dependencies?

thantos replied:
It does not! Been looking at that too.

Technically any of the keys could be requested at runtime event if they are not preWarmed. (and new entries can be added.)

I'll add a task to dynamically build the contents of the PreWarmContext instead of using an object.

Because the `aws_sdk` import is used, the impact of the extra methods is pretty minimal.

```js
function __f4() {
  return (function() {
    with({ aws_sdk_1: require("aws-sdk/lib/aws.js") }) {

return function /*lambda*/() {
    return this.getOrInit("LAMBDA", () => {
      var _a;

      return new aws_sdk_1.Lambda((_a = this.clientConfigRetriever) === null || _a === void 0 ? void 0 : _a.call(this, "LAMBDA"));
    });
  };

    }
  }).apply(undefined, undefined).apply(this, arguments);
}

function __f5() {
  return (function() {
    with({ aws_sdk_1: require("aws-sdk/lib/aws.js") }) {

return function /*eventBridge*/() {
    return this.getOrInit("EVENT_BRIDGE", () => {
      var _a;

      return new aws_sdk_1.EventBridge((_a = this.clientConfigRetriever) === null || _a === void 0 ? void 0 : _a.call(this, "EVENT_BRIDGE"));
    });
  };

    }
  }).apply(undefined, undefined).apply(this, arguments);
}
...
var __preWarmContext_proto = {};
Object.defineProperty(__f3, "prototype", { value: __preWarmContext_proto });
Object.defineProperty(__preWarmContext_proto, "constructor", { configurable: true, writable: true, value: __f3 });
Object.defineProperty(__preWarmContext_proto, "lambda", { configurable: true, writable: true, value: __f4 });
Object.defineProperty(__preWarmContext_proto, "eventBridge", { configurable: true, writable: true, value: __f5 });
Object.defineProperty(__preWarmContext_proto, "registerCustom", { configurable: true, writable: true, value: __f6 });
Object.defineProperty(__preWarmContext_proto, "custom", { configurable: true, writable: true, value: __f7 });
Object.defineProperty(__preWarmContext_proto, "getOrInit", { configurable: true, writable: true, value: __f8 });
Object.defineProperty(__preWarmContext_proto, "getOrFail", { configurable: true, writable: true, value: __f9 });
```

thantos replied:
Fixed this, new API doesn't serialize unless used.

sam-goodwin commented on src/function.ts:
```diff
@@ -61,28 +112,347 @@ export class Function<P, O> implements Integration {
+    super(_resource);
+
+    // retrieve and bootstrap all found native integrations. Will fail if the integration does not support native integration.
+    const nativeIntegrationsPrewarm = integrations
+      .map((i) => new IntegrationImpl(i))
+      .map((integration) => {
+        const integ = integration.native;
+        integ.bootstrap(this);
+        return integ;
+      })
+      .filter((i) => i.preWarm)
+      .map((i) => i.preWarm);
```
nit: I personally think this can be simplified.

```ts
const nativeIntegrationsPrewarm = integrations.flatMap(i => {
  const integ = new IntegrationImpl(i).native;
  integ.bootstrap(this);
  return integ.preWarm ? [integ] : [];
});
```

thantos replied:
nice :+1:

sam-goodwin commented on src/function.ts:
```diff
@@ -61,28 +112,347 @@ export class Function<P, O> implements Integration {
+      );
+  }
+
+  public static fromFunction<P, O>(
+    func: aws_lambda.IFunction
+  ): ImportedFunction<P, O> {
+    return new ImportedFunction<P, O>(func);
+  }
```
static functions above constructor

sam-goodwin commented on src/function.ts:
```diff
@@ -61,28 +112,347 @@ export class Function<P, O> implements Integration {
+  }
+}
+
+export class ImportedFunction<P, O> extends FunctionBase<P, O> {
```
docs

sam-goodwin commented on src/function.ts:
```diff
@@ -61,28 +112,347 @@ export class Function<P, O> implements Integration {
+  }
+}
+
+type ConditionalFunction<P, O> = P extends undefined
```
nice

sam-goodwin commented on src/function.ts:
```diff
@@ -61,28 +112,347 @@ export class Function<P, O> implements Integration {
+    this.scope = scope;
+    // Lets give the function something lightweight while we process the closure.
+    return aws_lambda.Code.fromInline(
+      "If you are seeing this in your lambda code, ensure generate is called and then consult the README."
```
Can we link to an issue?

thantos replied:
https://github.com/sam-goodwin/functionless/issues/128

sam-goodwin commented on src/function.ts:
```diff
@@ -61,28 +112,347 @@ export class Function<P, O> implements Integration {
+        isFactoryFunction: true,
+        serialize: (obj) => {
+          if (typeof obj === "string") {
+            const reversed =
+              Tokenization.reverse(obj, { failConcat: false }) ??
+              Tokenization.reverseString(obj).tokens;
+            if (!Array.isArray(reversed) || reversed.length > 0) {
+              if (Array.isArray(reversed)) {
+                tokens = [...tokens, ...reversed.map((s) => s.toString())];
+              } else {
+                tokens = [...tokens, reversed.toString()];
+              }
+            }
```
What's this do? Why do we need to reverse things?

thantos replied:
reverse ~= `de-encode`, and yeah, it helps us find real tokens.

sam-goodwin commented on src/function.ts:
```diff
@@ -61,28 +112,347 @@ export class Function<P, O> implements Integration {
+    );
+
+    const tokenContext = tokens.map((t) => {
+      const id = new RegExp("\\${Token\\[TOKEN\\.([0-9]*)\\]}", "g").exec(
```
nit: why not use native syntax?
```
/\${Token\..etc.../g
```

thantos replied:
We can, originally I was using variables, I guess I am no longer.

sam-goodwin commented on src/function.ts:
```diff
@@ -61,28 +112,347 @@ export class Function<P, O> implements Integration {
+    });
+
+    const resultText = tokenContext.reduce(
+      // TODO: support templated strings
```
i don't understand.

thantos replied:
I had assumed that `"${TOKEN[TOKEN_123]} and ${TOKEN[TOKEN_123]}"` would not work, but it appears like it does because the actual references are ripped out.

Before: `var __bus = {eventBusArn: "${TOKEN[TOKEN_439]}"};`
After: `var __bus = {eventBusArn: process.env.env__functionless439};`

So a templated string would look like:

`${_bus.eventBusArn} and ${_bus.eventBusArn}`

The problem is, what if a token is ever not in a standalone string?

`"${TOKEN[TOKEN_123]} stuff"`

Now the replace will not work. Need to figure out the best way to test this.

thantos replied:
I have not found a test that actually fails yet, so removed the comment.

sam-goodwin commented on src/function.ts:
```diff
@@ -61,28 +112,347 @@ export class Function<P, O> implements Integration {
+      result.text
+    );
+
+    // console.log(resultText);
```
remove

sam-goodwin commented on src/function.ts:
```diff
@@ -61,28 +112,347 @@ export class Function<P, O> implements Integration {
+
+    // console.log(resultText);
+
+    const asset = aws_lambda.Code.fromAsset("", {
```
does the empty string have any negative consequences?

thantos replied:
Only on hashing, but adding `user: ...` fixes that. Its stupid the input is required when the bundler can reference any path.

sam-goodwin commented on src/function.ts:
```diff
@@ -61,28 +112,347 @@ export class Function<P, O> implements Integration {
+  }
+}
+
+export interface NativeIntegration<F extends AnyFunction> {
```
docs

sam-goodwin commented on src/function.ts:
```diff
@@ -61,28 +112,347 @@ export class Function<P, O> implements Integration {
+  preWarm?: (preWarmContext: NativePreWarmContext) => void;
+}
+
+/* istanbul ignore next */
```
why do i keep seeing this istanbul thing everywhere?

Also, this class seems important - let's add docs.

thantos replied:
Removed them all, just means we might have slightly bigger bundles while testing and that is fine.

sam-goodwin commented on src/function.ts:
```diff
@@ -61,28 +112,347 @@ export class Function<P, O> implements Integration {
+    this.cache = {};
+  }
+
+  public lambda(): Lambda {
+    return this.getOrInit("LAMBDA", () => new Lambda());
+  }
+
+  public eventBridge(): EventBridge {
+    return this.getOrInit("EVENT_BRIDGE", () => new EventBridge());
+  }
```
I like how this class caches things, but is it a good idea to embed the individual clients in here? Won't that harm tree-shaking?

thantos replied:
Good question, answered above that the impact is low due to the use of `require("aws_sdk")`, but I would also like to reduce the size of the outputted code in any way...

The problem I had was... any integration may need any client, and I wouldn't want a client to add it's own flavor behind a shared key like `this.getOrInit("LAMBDA", () => new SpecialLambdaClient())`.

I guess we could make constants that encode the shared logic.

```ts
prewarmClient.get(PrewarmClient.LAMBDA);

// I am not sure how well an enum would serialize either
enum PrewarmClient {
   LAMBDA: () => new Lambda();
}
```

Will play with it.

thantos replied:
Got it

```ts
      preWarm: (preWarmContext: NativePreWarmContext) => {
        preWarmContext.getOrInit(PrewarmClients.LAMBDA);
      },
```

Serialized, only the required methods: 

```ts
var __exports_PrewarmClients = {};
var __exports_PrewarmClients_LAMBDA = {key: "LAMBDA", init: __f2};
__exports_PrewarmClients.LAMBDA = __exports_PrewarmClients_LAMBDA;
__exports.PrewarmClients = __exports_PrewarmClients;
```

sam-goodwin commented on src/function.ts:
```diff
@@ -61,28 +112,347 @@ export class Function<P, O> implements Integration {
+    return this.getOrFail<T>(key);
+  }
+
+  private getOrInit<T>(key: string, create: () => T): T {
```
Why not just make this public?

thantos replied:
I probably will now after the last comment. Original thought was that shared keys should use the known helpers and the custom keys should try to use `register` in the preWarm and `get` in the runtime.

sam-goodwin commented on src/function.ts:
```diff
@@ -61,28 +112,347 @@ export class Function<P, O> implements Integration {
+    return this.getOrInit("EVENT_BRIDGE", () => new EventBridge());
+  }
+
+  public registerCustom<T>(
+    key: string,
+    init: () => T,
+    allowExists: boolean = false
+  ): T {
+    if (key in this.cache && !allowExists) {
+      throw Error(
+        `Prewarm Context key ${key} already exists. Use context.custom to get the object or set allowExists to true.`
+      );
+    }
+    return this.getOrInit<T>(key, init);
+  }
```
Is this over-engineering? Seems like `getOrInit` and `getOrFail` are enough?

thantos replied:
Maybe! I liked being able to assert that a key already existed, but I guess we could just add a `get: () => T | null`.

sam-goodwin commented on test-app/src/message-board.ts:
```diff
@@ -376,15 +371,11 @@ interface TestDeleteEvent
-    runtime: aws_lambda.Runtime.NODEJS_14_X,
-    handler: "index.handler",
-  })
+  stack,
+  "sendNotification",
+  async (event) => {
+    console.log("notification: ", event);
+  }
 );
```
syntax highlighting is broken by the string literal types in the Comment interface - consider moving them to the bottom to aid legibility in GitHub

sam-goodwin commented on test/function.localstack.test.ts:
```diff
@@ -0,0 +1,332 @@
+        return new Function(parent, "function", async () => val);
+      };
+
+      return { outputs: { function: create().resource.functionName } };
+    },
+    async (context) => {
+      await testFunction(context.function, {}, "a");
```
This seems boiler-platey and confusing. Can we simplify somehow?

```ts
const func = new Function(.., (param1, param2) => ..);

expect (await synth(func)(param1, param2)).toEqual(result);
```

thantos replied:
Was trying to reduce the number of stack deployments, it also supports automatic creation of outputs and passes them back to the test function. Deployments to `localstack` are not fast and would require either a unique state, or for the same stack to incrementally deploy during each test. 

The `outputs: { function: ... }` part is boilerplaty, but this wrapper supports any resource (there is example that uses the bus arns below). 

I'll see if I can make a special wrapper for the lambda tests to make them smaller.

What does `synth` do here?

thantos replied:
refactored

```ts
testFunctionResource("name", () => new Function(...), expected, payload?);
```

sam-goodwin commented on test/function.localstack.test.ts:
```diff
@@ -0,0 +1,332 @@
+    (parent) => {
+      const bus = new EventBus<any>(parent, "bus");
+
+      // Necessary to keep the bundle small and stop the test from failing.
```
How significantly does this effect the bundle size?

thantos replied:
3mb vs 100kb (including the serialized coverage code).

sam-goodwin replied:
Can we fix this now that we have control over the closure serializer?

sam-goodwin commented on test/function.localstack.test.ts:
```diff
@@ -0,0 +1,332 @@
+      const bus = new EventBus<any>(parent, "bus");
+
+      // Necessary to keep the bundle small and stop the test from failing.
+      // See https://github.com/sam-goodwin/functionless/pull/103#issuecomment-1116396779
```
Should we link to an issue instead of a comment?

sam-goodwin commented on test/function.localstack.test.ts:
```diff
@@ -0,0 +1,332 @@
+    }
+  );
+
+  // See https://github.com/sam-goodwin/functionless/pull/103#issuecomment-1116396779
```
Issue instead of comment?

sam-goodwin commented on test/function.localstack.test.ts:
```diff
@@ -0,0 +1,332 @@
+
+  // See https://github.com/sam-goodwin/functionless/pull/103#issuecomment-1116396779
+  testResource.skip(
+    "Call Lambda AWS SDK put event to bus without reference",
+    (parent) => {
+      const bus = new EventBus<EventBusRuleInput>(parent, "bus");
+
+      const func = new Function(parent, "function", async () => {
+        const result = $AWS.EventBridge.putEvents({
```
Is this really not supported?

thantos replied:
It is supported, but it breaks in jest.

Essentially, pulumi serializes the entire `aws.js` file, which has `jest`'s injection. Then `global[ts-jest]` doesn't exist and fails in the lambda when its run on localstack.

Doing this outside of tests works, but does bloat the output bundle. #122 is supposed to reduce the change of this case.

sam-goodwin commented:
How would someone write unit tests for their Lambda Functions that are coded this way? Do they have to mock much? Is there a pattern we can encourage?

thantos commented:
> How would someone write unit tests for their Lambda Functions that are coded this way? Do they have to mock much? Is there a pattern we can encourage?

Good question, right now there isn't an easy way. We could just expose the closure through the function and let them call it. Obviously it would have the problem that it assumes the serialization was lossless. Or we could serialize the function, import it again (or evaluate it) and then call the serialized from.


```ts
new Function(..., () => {}).invoke();
```

sam-goodwin commented on src/function.ts:
````diff
@@ -1,43 +1,94 @@
- * ```
- */
-export class Function<P, O> implements Integration {
+export type FunctionClosure<P, O> = (payload: P) => Promise<O>;
````
We should pass the Lambda Context from `@types/aws-lambda` as the second argument

sam-goodwin commented:
Is the following supported?

```ts
let _allowList;
const loadAllowList = async () => {
  return (_allowList =
    _allowList ?? (await fs.promises.readFile("allow-list.json")));
};

new Function(scope, "foo", async () => {
  // reference the allowList here instead
  const list = await loadAllowList();
});
```

Will the `allow-list.json` file be bundled? Probably not right?

thantos commented:
Ran into issues that were difficult to solve with the current state of Pulumi's serializer. Working on a fork of pulumi:

Added features so far:

* Symbol support - Symbols cause an error when serialized - https://github.com/pulumi/pulumi/issues/6229#issuecomment-1135118712
* Ability to update the object being serialized (see below)

Updating Serialized objects:

Current Pulumi supports a `serialize` callback which allows consumers to ignore objects during serialization. The reality of this is entire objects can be ignored (turned into `undefined`), but if the object is partially needed by the runtime, then its all or nothing.

For example, a CDK Construct like a `Table` can be discovered and ignored:

```ts
if(obj instanceof aws_dynamo.Table) {
   return false;
}
```

But this fails when we need parts of the table

```ts
new Function(async () => {
    const func = (table) => table.tableName;
    console.log(func(table));
})
```

In the above case, returning false would cause `table` to be undefined and returning true would cause the entire CDK `Table` to be serialized, generally failing when a `Lazy` is found in the CDK object. Serializing the CDK construct also causes a huge bloat to the closure that is generated. I've seen simple code generate a 20MB bundle.

There are ways to work around the issue above, for example setting the table name outside the closure to avoid the entire table being serialized.

```ts
const tableName = table.tableName;
new Function(async () => {
    const func = (tableName) => tableName; // not quite the same...
    console.log(func(tableName));
})
```

While generally a nuisance that can be worked around, this created an issue when trying integrate dynamo with the `$AWS.DynamoDB.*` operations. 

```ts
const table = new Table();
new Function(async () => {
    return $AWS.DynamoDB.getItem({ TableName: table, ... }); // will also serialize the entire table.
})
```

Working on a base assumption that there are parts of the CDK object we should not need (and if we do, we should add escape hatches to unblock), we are adding a new feature to our pulumi serializer fork: The ability to override the object being serialized (instead of ignoring it).

1. Updated the `serialize` contract to accept `boolean | any`: (o: any) => boolean | any;
2. When a non-boolean truthy value is returned, use that value instead.

Now we can do

```ts
if (Resource.isResource(o as any)) {
    const { node, stack, env, ...rest } = o as unknown as Resource;
    return rest;
}
```

This block removes node, stack, and env from the object to be serialized, but leaves the rest intact.

What we are modifying so far:
* CDK Constructs/Resources - stack, node, env are not needed at least
* Cfn* L1 resources - stack, node, creationStack, _toCloudformation, and cfnProperties are not needed.
* CDK ITaggable - `tags` on constructs can easily bring in the entire CDK TagManager which appears to have minimal values at tuneimt.
* Functionless Integrations - Functionless integrations (which can also be or contain CDK constructs) can contain integrations for any of the functionless resources (step functions, app sync, etc), but we only need the `native` functions, in fact, we only need `preWarm` and `call` as bind is used during `synth`.
* Table - ?

sam-goodwin commented:
> Updated the serialize contract to accept boolean | any: (o: any) => boolean | any;

This has an edge case - what happens when the actual value (we want to keep) is a literal false boolean?

thantos commented:
> > Updated the serialize contract to accept boolean | any: (o: any) => boolean | any;
> 
> This has an edge case - what happens when the actual value (we want to keep) is a literal false boolean?

I don't like the current contract either, the current form was the fast to create to experiment.... however... you example isn't a problem as we don't overwrite then the value is a boolean

https://github.com/functionless/nodejs-closure-serializer/blob/main/src/closure/createClosure.ts#L839

The real edge case is if I wanted to update any value to a boolean (which I figured was a rare case for now...). 

* false - updates to undefined
* true - leaves as is
* anything else - updates to that value

Proposed update: https://github.com/functionless/nodejs-closure-serializer/issues/1

thantos commented:
From review today: Moved the functionProps to an optional 3rd param instead of last in `Function`

sam-goodwin commented on src/function.ts:
```diff
@@ -61,28 +125,539 @@ export class Function<P, O> implements Integration {
+             */
+            const transformCfnResource = (o: unknown): any => {
+              if (Resource.isResource(o as any)) {
+                const { node, stack, env, ...rest } = o as unknown as Resource;
+                return rest;
```
cute! does this capture methods from the prototype?

sam-goodwin replied:
How big are the bundles we're getting?

thantos replied:
that large lambda I have in message-board is 200kb. See attached.
[index.txt](https://github.com/functionless/functionless/files/8791814/index.txt)

sam-goodwin commented on src/function.ts:
```diff
@@ -61,28 +125,539 @@ export class Function<P, O> implements Integration {
+              } else if (Token.isUnresolved(o)) {
+                const token = (<any>o).toString();
+                // add to tokens to be turned into env variables.
+                tokens = [...tokens, token];
```
why not push?

thantos replied:
I don't like using the mutation functions on arrays if avoidable.

sam-goodwin replied:
But you're setting a variable. So still mutational at the cost of o(n) time

sam-goodwin commented on src/function.ts:
```diff
@@ -61,28 +125,539 @@ export class Function<P, O> implements Integration {
+                o.cfnResourceType ===
+                aws_dynamodb.CfnTable.CFN_RESOURCE_TYPE_NAME
+              ) {
+                const table = o as aws_dynamodb.CfnTable;
+                if (!table.streamSpecification) {
+                  const { attrStreamArn, ...rest } = table;
+
+                  return rest as unknown as CfnResource;
+                }
```
What's the logic here? Seems like we're keeping a whole bunch

thantos replied:
the table is also going through the CfnResource filtering transform, this just adds the safety for the streamArm.

thantos replied:
Updated the docs.

sam-goodwin commented on src/function.ts:
```diff
@@ -61,28 +125,539 @@ export class Function<P, O> implements Integration {
+              if (o && typeof o === "object" && "kind" in o) {
+                const integ = o as Integration;
+                const {
+                  appSyncVtl,
+                  asl,
+                  native: { call, preWarm } = {},
```
We have to remember to update this every time we add a new integration type. That's probably fine because it won't happen often, right?

thantos replied:
We don't NEED to, but new integration types may bloat the closures. I'll make this more dynamic...

thantos replied:
Some typesafe runtime enforcement

 ```ts
/**
 * Maintain a typesafe runtime map of integration type keys to use elsewhere.
 * 
 * For example, removing all but native integration from the {@link Function} closure.
 */
const INTEGRATION_TYPES: Record<keyof IntegrationMethods<any>, any> = {
  appSyncVtl: true,
  asl: true,
  native: true
} 

export const INTEGRATION_TYPE_KEYS = Object.keys(INTEGRATION_TYPES) as (keyof IntegrationMethods<any>)[];
```

thantos replied:
slightly better

```ts
/**
 * Maintain a typesafe runtime map of integration type keys to use elsewhere.
 *
 * For example, removing all but native integration from the {@link Function} closure.
 */
const INTEGRATION_TYPES: { [P in keyof IntegrationMethods<any>]: P } = {
  appSyncVtl: "appSyncVtl",
  asl: "asl",
  native: "native",
};

export const INTEGRATION_TYPE_KEYS = Object.values(INTEGRATION_TYPES);
```

sam-goodwin commented on src/function.ts:
```diff
@@ -61,28 +125,539 @@ export class Function<P, O> implements Integration {
+      result.text
+    );
+
+    // // find any strings that contain tokens
+    // const stringsWithQuotes = /["'`].*\${Token\[.*\.[0-9]*\]}.*['`"]/g.exec(
+    //   resultText
+    // );
+
+    // stringsWithQuotes?.reduce(
+    //   (r, s) => r.replace(s, s.replace(/(^['"]|['"]$)/g, "`")),
+    //   resultText
+    // );
```
nit: remove?

sam-goodwin commented on src/function.ts:
```diff
@@ -61,28 +125,539 @@ export class Function<P, O> implements Integration {
+  LAMBDA: {
+    key: "LAMBDA",
+    init: (key, props) =>
+      new (require("aws-sdk").Lambda)(props?.clientConfigRetriever?.(key)),
```
`aws-sdk` is ideal for bundling because it's included in the lambda function by default. Do we know if it has any performance implications though? Does V2 use the optimized C code that v3 does?

thantos replied:
There may be some performance improvements using v3 in a function: https://aws.amazon.com/blogs/developer/modular-packages-in-aws-sdk-for-javascript/

https://github.com/functionless/functionless/issues/174

sam-goodwin commented on test-app/src/message-board.ts:
```diff
@@ -425,6 +398,88 @@ customDeleteBus
+        id: "from the test method!!",
+      },
+    });
+    // busbus({
+    //   "detail-type": "Delete-Post-Success",
+    //   source: "MessageDeleter",
+    //   detail: {
+    //     id: "from the test method!!",
+    //   },
+    // });
```
remove

sam-goodwin commented on test-app/src/message-board.ts:
```diff
@@ -425,6 +398,88 @@ customDeleteBus
+  stack,
+  "testFunc",
+  async () => {
+    console.log(customDeleteBus.eventBusArn);
+    console.log(busbusbus.eventBusArn);
+    console.log("huh?!?!?!??!!");
+    console.log("strange");
+    const result = func();
+    console.log(`function result: ${result}`);
```
remove logging?

sam-goodwin commented on test/function.localstack.test.ts:
```diff
@@ -0,0 +1,533 @@
+      },
+      async (context) => {
+        const exp =
+          // @ts-ignore
+          typeof expected === "function" ? expected(context) : expected;
+        // @ts-ignore
```
why ignore?

thantos replied:
Laziness mostly

![image](https://user-images.githubusercontent.com/289213/170843103-7674d65b-d6d6-45ac-8d9d-a7aaee44b3d8.png)

sam-goodwin commented on test/function.localstack.test.ts:
```diff
@@ -0,0 +1,533 @@
+    return new Function(parent, id, async () => val);
+  };
+
+  testFunctionResource(
+    "Call Lambda from closure with parameter multiple 1",
+    (parent) => create(parent, "func6", "c"),
+    "c"
+  );
+
+  testFunctionResource(
+    "Call Lambda from closure with parameter multiple 2",
+    (parent) => create(parent, "func7", "d"),
+    "d"
+  );
```
multiple 1 and 2. What does that mean?

sam-goodwin commented on test/function.localstack.test.ts:
```diff
@@ -0,0 +1,533 @@
+
+  testFunctionResource(
+    "Call Lambda payload",
+    (parent) =>
+      new Function(parent, "function", async (event: { val: string }) => {
+        return `value: ${event.val}`;
+      }),
+    "value: hi",
+    { val: "hi" }
```
so just so i understand, the contract is:
1. create the Function construct
2. expected output
3. input argument

sam-goodwin replied:
For users, we're going to need this to be:
```ts
test("blah", async () => {
  const func = new Function(..);
  expect(await func(input)).toEqual(expected);
});
```

thantos replied:
#131 covers running the closure of integrations as typescript.

The tests here look funky because of the integration with localstack and actually running the lambda on localstack. This is how we'd test any CDK resource, not specific to our function.

Did you have an idea how testing would be done by developer? I don't think we'd want to integrate localstack in official, so its up to the developer how they'd test resources running again AWS or localstack (which look like any other CDK resource).

I guess #131 is the closest to what developers can do today, just running their lambda code, but without the step of serialization and integration being the unknown variables. 

Another option is to expose the serialized form somehow to run locally to catch any issues... not sure yet.

sam-goodwin commented on test/function.localstack.test.ts:
```diff
@@ -0,0 +1,533 @@
+    0
+  );
+
+  // See https://github.com/sam-goodwin/functionless/pull/122
```
fix link

thantos replied:
Fixed all of the previous links in #175.

Fixed all of the links from this change in the next commit.

thantos commented on .vscode/launch.json:
```diff
@@ -1,6 +1,21 @@
 {
   "version": "0.2.0",
   "configurations": [
+    {
+      "type": "aws-sam",
+      "request": "direct-invoke",
+      "name": "message-board-testFuncB346065B-zllkEyeNxDAI:index.handler (nodejs14.x)",
+      "invokeTarget": {
+        "target": "code",
+        "projectRoot": "${workspaceFolder}/message-board-testFuncB346065B-zllkEyeNxDAI",
+        "lambdaHandler": "index.handler"
+      },
+      "lambda": {
+        "runtime": "nodejs14.x",
+        "payload": {},
+        "environmentVariables": {}
+      }
+    },
```
Probably not

thantos commented on src/aws.ts:
```diff
@@ -361,4 +601,37 @@ export namespace $AWS {
+      "EventBridge.putEvent"
+    >({
+      kind: "EventBridge.putEvent",
+      native: {
```
Not sure what you mean?

sam-goodwin replied:
Can we code generate `native` integrations for the entire AWS SDK?

thantos replied:
ohh, yeah, we could. https://github.com/functionless/functionless/issues/168

thantos commented on src/compile.ts:
```diff
@@ -269,6 +287,54 @@ export function compile(
+            node.expression,
+            Function.FunctionlessType
+          ) &&
+          // only take the form with the arrow function at the end.
+          (node.arguments?.length === 3 || node.arguments?.length === 4) &&
+          ts.isArrowFunction(node.arguments[2])
```
Yes. WIP.

thantos commented on src/compile.ts:
```diff
@@ -269,6 +287,54 @@ export function compile(
+      }
+
+      /**
+       * Heuristically evaluate the fqn of a symbol to be in a module and of a type name.
+       *
+       * /somePath/node_modules/{module}/somePath.{typeName}
```
Any other thoughts for handling this heuristically?

sam-goodwin replied:
Structural duck typing. Check for the existence of properties.

sam-goodwin replied:
See https://github.com/functionless/functionless/pull/103#discussion_r884030512

thantos replied:
What property would you look for on the base CDK construct?

sam-goodwin replied:
I would check for the node property and then check for many of the properties on the node class.

thantos replied:
Actually, this pattern works for both nested and top level paths (though the comment doesn't mention that).

All it does is check to see if the npm package name and the type name are equal.

Both will return true

`/home/sussmans/functionless/node_modules/constructs/lib/construct.Construct`
and 
`/home/sussmans/functionless/node_modules/aws-cdk-lib/node_modules/constructs/lib/construct.Construct`

I think this will work for moth situations.

sam-goodwin replied:
Cool, sounds good.

thantos commented on src/compile.ts:
```diff
@@ -856,6 +1175,167 @@ export function compile(
+       *
+       * The call to busbus can be resolved to bus if the scope is the array function.
+       */
+      function getOutOfScopeValueNode(
```
Probably cannot be found. Have not exhaustively tested external reference patterns. Good test to add.

sam-goodwin replied:
Let's open an issue or track in a test?

thantos replied:
got it :-) (commit soon)

```ts
const b = {
    bus: new EventBus(...)
}
new Function(... () => {
    const { bus } = b;
    bus(...);
})
```

```ts
{ integration: b["bus"] }
```

thantos replied:
summary: created a method to discover and flatten destructured assignment.

```ts
      function flattenDestructuredAssignment(
        element: ts.BindingElement
      ): ts.ElementAccessExpression {
        // if the binding renames the property, get the original
        // { a : x } -> a
        // { a } -> a
        // [a] -> 0
        const name = ts.isArrayBindingPattern(element.parent)
          ? element.pos
          : // binding renames the property or is a nested binding pattern.
          element.propertyName
          ? element.propertyName
          : // the "name" can be a binding pattern. In that case the propertyName will be set.
            (element.name as ts.Identifier);

        const getParent = () => {
          // { a } = b;
          if (ts.isVariableDeclaration(element.parent.parent)) {
            if (!element.parent.parent.initializer) {
              throw Error(
                "Expected a initializer on a destructured assignment: " +
                  element.getText()
              );
            }
            return element.parent.parent.initializer;
          } else if (ts.isBindingElement(element.parent.parent)) {
            return flattenDestructuredAssignment(element.parent.parent);
          } else {
            throw Error(
              "Cannot flatten destructured parameter: " + element.getText()
            );
          }
        };

        const parent = getParent();

        // always use element access as this will work for all possible values.
        // [parent][name]
        return ts.factory.createElementAccessExpression(
          parent,
          typeof name !== "number"
            ? ts.isComputedPropertyName(name)
              ? name.expression
              : ts.isIdentifier(name)
              ? // { a } = b -> b["a"]
                ts.factory.createStringLiteral(name.text)
              : name
            : // numeric index position for a destructured array
              name
        );
      }
```

sam-goodwin replied:
So the synth-time interpreters will never have to handle restructuring? Is there any chance of name conflicts after flattening?

thantos replied:
We are not always flattening. I am only doing it for getting the "out of scope" version of the integration to be able to pass back into the function. In this case at least, the names won't collide.

thantos commented on src/compile.ts:
```diff
@@ -407,6 +473,261 @@ export function compile(
+            integrations.push({ expr: integ, args }),
+        };
+
+        const body = toNativeExpr(
```
Yeah

thantos commented on src/compile.ts:
```diff
@@ -407,6 +473,261 @@ export function compile(
+            // call the integration call function with the prewarm context and arguments
+            // At this point, we know native will not be undefined
+            // await integration.native.call(args, preWarmContext)
+            // FIXME: doesn't work without an array
```
Woops. Missed removing this.

thantos commented on src/event-bridge/event-bus.ts:
```diff
@@ -120,9 +129,39 @@ abstract class EventBusBase<E extends EventBusRuleInput>
+              DetailType: event["detail-type"],
+              Resources: event.resources,
+              Source: event.source,
+              Time: event.time ? new Date(event.time) : undefined,
```
Think the timestamp will ever be 0?

sam-goodwin replied:
Not sure, but it is a valid timestamp. Easy enough to check:

```ts
Time: typeof event.time === "number" ? new Date(event.time) : undefined
```

sam-goodwin commented on src/compile.ts:
```diff
@@ -269,6 +287,54 @@ export function compile(
+        return mod === module && type === typeName;
+      }
+
+      function isCDKConstruct(type: ts.Type): boolean {
+        const typeSymbol = type.getSymbol();
+
+        return (
+          ((typeSymbol &&
+            isInstanceOf(typeSymbol, "constructs", "Construct")) ||
+            type.getBaseTypes()?.some((t) => isCDKConstruct(t))) ??
+          false
+        );
+      }
```
I would check for the node property and then check for many of the properties on the node class.

thantos replied:
Answered here: https://github.com/functionless/functionless/pull/103#discussion_r884063485

I think this will work. The chance that a package name and type name collide in the same code base is low.

thantos commented on src/compile.ts:
```diff
@@ -45,6 +46,12 @@ export function compile(
+        return sf;
+      }
+
+      // TODO: this seems like a pattern, make a "public import context"
```
I changed compile to only include the functionless import when its used, instead of all of the time. 

Originally thought we'd have multiple dynamic imports, but ended up removing the other one. Removed the comment.

thantos commented on src/compile.ts:
```diff
@@ -407,6 +473,261 @@ export function compile(
+       *            // call can make use of the cached clients in prewarmContext to avoid duplicate inline effect
+       *            await bus.native.call(prewarmContext, { source: "src" });
+       *        },
+       *     [bus] // integrations
```
`table.get`

The entire expression on the call is passed up or corrected to be the out of scope form.

```ts
const table = new Table();
new Function(async () => {
   const t  = table;
   const get = t.get;
   get(...);
})
```

becomes

`[table.get]`

sam-goodwin replied:
That is awesome. Great job

thantos commented on src/compile.ts:
```diff
@@ -407,6 +473,261 @@ export function compile(
+            );
+
+            if (!outOfScopeIntegrationReference) {
+              // integration variables can be CDK constructs which will fail serialization.
+              // TODO: can we relax this be determining if a integration reference will fail?
```
Removed it

thantos commented on src/function.ts:
```diff
@@ -1,43 +1,107 @@
-    // Integration object for appsync VTL
+    const functionName = this.resource.functionName;
+
+    // Native is used when this function is called from another lambda function's serialized closure.
+    // define this here to closure the functionName without using `this`
```
Trying this... maybe we need a lazy?

```ts
let func1: Function<string, string> | undefined;
const func1Lazy = Lazy.integ(() => func1);
const func2 = new Function(async () => {
   func1Lazy.get()("hello")
})
func1 = new Function<string, string>(async () =>{ func2() });
```

This won't work right now because we invoke the integration in the constructor of the Function its used in. 

Or maybe a ref()

```ts
const func1Ref = Function.ref<string, string>();
const funct2 = new Function(async () => {
   func1Ref("");
})
const func1 = func1Ref.set(new Function(async () => { func2() }));
```

Right now... this would just mean calling the bind function and re-writing the environment variables, I think...

https://github.com/functionless/functionless/issues/173

thantos commented on src/function.ts:
```diff
@@ -61,28 +125,539 @@ export class Function<P, O> implements Integration {
+    );
+
+    // Start serializing process, add the callback to the promises so we can later ensure completion
+    callbackLambdaCode &&
+      Function.promises.push(
+        callbackLambdaCode.generate(nativeIntegrationsPrewarm)
+      );
+  }
```
FINE :-P

thantos commented:
> Can we fix this now that we have control over the closure serializer?

No idea why github isn't showing the repy box...

This case works for users now (message-board has an example).

But it doesn't work in our tests because all of $AWS is being embedded which has been including the `globals["ts-jest"]` reference which when breaks all tests. May be able to play with `serialize` function to change the behavior.
